hconf.c

```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
         return (0);
 }
 
+static int
+hconf_get_switch(struct hconf_softc *sc, int swtype, u_int *mask)
+{
+	uint8_t *fbuf;
+	int error;
+
+	if (sc->switch_rlen[swtype] <= 1)
+		return (ENXIO);
+
+	fbuf = malloc(sc->switch_rlen[swtype], M_TEMP, M_WAITOK | M_ZERO);
+	sx_xlock(&sc->lock);
+
+	error = hid_get_report(sc->dev, fbuf, sc->switch_rlen[swtype], NULL,
+	    HID_FEATURE_REPORT, sc->switch_rid[swtype]);
+	if (error == 0) {
+		*mask = hid_get_data_unsigned(fbuf + 1,
+		    sc->switch_rlen[swtype] - 1, &sc->switch_loc[swtype]);
+	}
+
+	sx_unlock(&sc->lock);
+	free(fbuf, M_TEMP);
+	return (error);
+}
+
+static int
+hconf_set_switch(struct hconf_softc *sc, int swtype, u_int mask)
+{
+	uint8_t *fbuf;
+	int error;
+
+	if (sc->switch_rlen[swtype] <= 1)
+		return (ENXIO);
+
+	fbuf = malloc(sc->switch_rlen[swtype], M_TEMP, M_WAITOK | M_ZERO);
+	sx_xlock(&sc->lock);
+
+	error = hid_get_report(sc->dev, fbuf, sc->switch_rlen[swtype],
+	    NULL, HID_FEATURE_REPORT, sc->switch_rid[swtype]);
+	if (error != 0)
+		goto out;
+
+	hid_put_data_unsigned(fbuf + 1, sc->switch_rlen[swtype] - 1,
+	    &sc->switch_loc[swtype], mask);
+	error = hid_set_report(sc->dev, fbuf, sc->switch_rlen[swtype],
+	    HID_FEATURE_REPORT, sc->switch_rid[swtype]);
+
+out:
+	sx_unlock(&sc->lock);
+	free(fbuf, M_TEMP);
+	return (error);
+}
+
+static int
+hconf_switch_handler(SYSCTL_HANDLER_ARGS)
+{
+	struct hconf_softc *sc = arg1;
+	u_int value;
+	int error;
+
+	error = hconf_get_switch(sc, arg2, &value);
```

Comment:
Microsoft does not require configuration page feature reports to be readable:
https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/touchpad-required-hid-top-level-collections
This means that old values should be taken from soft context rather then from hardware. hconf_get_switch() can be used at device_attach() handler to initialize variables.